txpool/txpool.go

```diff
@@ -823,466 +531,508 @@ func (t *TxPool) validateTx(
+	p.index.remove(pruned...)
+	p.gauge.decrease(slotsRequired(pruned...))
 
-// Remove removes the transaction with given hash
-func (t *txHeapWrapper) Remove(hash types.Hash) bool {
-	for i, tx := range t.txs {
-		if tx.Hash == hash {
-			t.txs = append(t.txs[:i], t.txs[i+1:]...)
-			return true
-		}
+	// update next nonce
+	account.setNonce(nonce)
+
+	if first := account.enqueued.peek(); first != nil &&
+		first.Nonce == nonce {
+		// first enqueued tx is expected -> signal promotion
+		p.promoteReqCh <- promoteRequest{account: addr}
 	}
-	return false
 }
 
-// Account based heap implementation //
-// The heap is min-nonce ordered //
+// createAccountOnce is used when discovering an address
+// of a received transaction for the first time.
+// This function ensures that the account
+// is created atomically and only once.
+func (p *TxPool) createAccountOnce(newAddr types.Address) *account {
+	// fetch nonce from state
+	stateRoot := p.store.Header().StateRoot
+	stateNonce := p.store.GetNonce(stateRoot, newAddr)
 
-type txHeap []*types.Transaction
+	// initialize the account
+	account := p.accounts.initOnce(newAddr, stateNonce)
 
-// Required method definitions for the standard golang heap package
+	return account
+}
 
-func (t *txHeap) Peek() *types.Transaction {
-	if len(*t) == 0 {
-		return nil
+/* QUERY methods */
+
+// GetNonce returns the next nonce for the account
+// -> Returns the value from the TxPool if the account is initialized in-memory
+// -> Returns the value from the world state otherwise
+func (p *TxPool) GetNonce(addr types.Address) uint64 {
+	account := p.accounts.from(addr)
+	if account == nil {
+		stateRoot := p.store.Header().StateRoot
+		stateNonce := p.store.GetNonce(stateRoot, addr)
+
+		return stateNonce
 	}
-	return (*t)[0]
-}
 
-func (t *txHeap) Len() int {
-	return len(*t)
+	return account.getNonce()
 }
 
-func (t *txHeap) Swap(i, j int) {
-	(*t)[i], (*t)[j] = (*t)[j], (*t)[i]
+// GetCapacity returns the current number of slots
+// occupied in the pool as well as the max limit
+func (p *TxPool) GetCapacity() (uint64, uint64) {
+	return p.gauge.read(), p.gauge.max
 }
 
-func (t *txHeap) Less(i, j int) bool {
-	return (*t)[i].Nonce < (*t)[j].Nonce
+// GetPendingTx returns the transaction by hash in the TxPool (pending txn) [Thread-safe]
+func (p *TxPool) GetPendingTx(txHash types.Hash) (*types.Transaction, bool) {
+	tx, ok := p.index.load(txHash)
+	if !ok {
+		return nil, false
+	}
+
+	return tx, true
 }
 
-func (t *txHeap) Push(x interface{}) {
-	(*t) = append((*t), x.(*types.Transaction))
+// GetTxs gets pending and queued transactions
+func (p *TxPool) GetTxs(inclQueued bool) (
+	allPromoted, allEnqueued map[types.Address][]*types.Transaction,
+) {
+	allPromoted, allEnqueued = p.accounts.allTxs(inclQueued)
+	return
 }
 
-func (t *txHeap) Pop() interface{} {
-	old := *t
-	n := len(old)
-	x := old[n-1]
-	*t = old[0 : n-1]
-	return x
+// Thread safe map of all accounts registered by the pool.
+type accountsMap struct {
```

Comment:
Feel free to move all accountsMap structs & account structs methods to a separate file: account.go

Reply:
[Exported](https://github.com/0xPolygon/polygon-sdk/pull/287/commits/8f8c2a793b02c4744e33f3598b871ea3b1aa73e9)